// app-modernization-mcp-server/dist/tools/ToolNames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsbenchToolNames = exports.UpgradeToolNames = exports.VscodeToolNames = exports.McpToolNames = void 0;
var McpToolNames;
(function (McpToolNames) {
    McpToolNames["JAVA_BUILD"] = "appmod-build-project";
    McpToolNames["JAVA_RUN_TEST"] = "appmod-run-test";
    McpToolNames["JAVA_FIX_TEST"] = "appmod-fix-test";
    McpToolNames["SEARCH_KB"] = "appmod-search-knowledgebase";
    McpToolNames["SEARCH_FILE"] = "appmod-search-file";
    McpToolNames["GET_KB"] = "appmod-fetch-knowledgebase";
    McpToolNames["MIGRATION_SUMMARY"] = "appmod-create-migration-summary";
    McpToolNames["RUN_TASK"] = "appmod-run-task";
    McpToolNames["CONSISTENCY_VALIDATE"] = "appmod-consistency-validation";
    McpToolNames["COMPLETENESS_VALIDATE"] = "appmod-completeness-validation";
    McpToolNames["INSTALL_APPCAT"] = "appmod-install-appcat";
    McpToolNames["KB_ASSESSMENT"] = "appmod-kb-assessment";
    McpToolNames["RUN_ASSESSMENT"] = "appmod-run-assessment";
    McpToolNames["PRECHECK_ASSESSMENT"] = "appmod-precheck-assessment";
    McpToolNames["VERSION_CONTROL"] = "appmod-version-control";
})(McpToolNames || (exports.McpToolNames = McpToolNames = {}));
// tool names in vs code extension
var VscodeToolNames;
(function (VscodeToolNames) {
    VscodeToolNames["CveFixTool"] = "appmod-validate-cve";
    VscodeToolNames["PreviewMarkdown"] = "appmod-preview-markdown";
    VscodeToolNames["GetConfigTool"] = "appmod-get-vscode-config";
    VscodeToolNames["OpenAssessmentReport"] = "migration_assessmentReport";
})(VscodeToolNames || (exports.VscodeToolNames = VscodeToolNames = {}));
// Java Upgrade tool names for Java build and environment management
var UpgradeToolNames;
(function (UpgradeToolNames) {
    UpgradeToolNames["InstallJdkTool"] = "install_jdk_for_java";
    UpgradeToolNames["InstallMavenTool"] = "install_maven_for_java";
    UpgradeToolNames["ListJdks"] = "list_jdks";
    UpgradeToolNames["ListMavens"] = "list_mavens";
    UpgradeToolNames["CveValidateTool"] = "validate_cves_for_java";
    UpgradeToolNames["BuildJavaProject"] = "build_java_project";
    UpgradeToolNames["RunTestsForJava"] = "run_tests_for_java";
})(UpgradeToolNames || (exports.UpgradeToolNames = UpgradeToolNames = {}));
// Below tools are used for msbench only
var MsbenchToolNames;
(function (MsbenchToolNames) {
    MsbenchToolNames["RunInTerminal"] = "appmod_run_in_terminal";
})(MsbenchToolNames || (exports.MsbenchToolNames = MsbenchToolNames = {}));
