// app-modernization-mcp-server/dist/core/request/RunTaskRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runTaskResponseSchema = exports.validationStage = exports.runTaskSchema = void 0;
const zod_1 = require("zod");
exports.runTaskSchema = zod_1.z.object({
    kbId: zod_1.z.string().optional().describe('The ID used to retrieve the knowledge base'),
    scenario: zod_1.z.string().optional().describe('The scenario for the migration'),
    workspacePath: zod_1.z.string().describe('The workspace path for source code to be migrated'),
});
exports.validationStage = zod_1.z.object({
    title: zod_1.z.string().describe('The title of the validation stage'),
    instruction: zod_1.z.string().describe('The instruction for the validation stage'),
});
exports.runTaskResponseSchema = zod_1.z.object({
    sessionId: zod_1.z.string().describe('The session ID for the migration task'),
    runMode: zod_1.z
        .enum(['interactive', 'non-interactive'])
        .describe('The run mode for the migration task'),
    goal: zod_1.z.string().describe('The goal for the migration task'),
    scope: zod_1.z.string().describe('The scope for the migration task'),
    successCriteria: zod_1.z.string().describe('The success criteria for the migration task'),
    toolUsageInstructions: zod_1.z
        .string()
        .describe('The tool usage instructions for the migration task'),
    progressTrackingInstructions: zod_1.z.string().describe('The progress tracking instructions'),
    planningInstructions: zod_1.z.string().describe('The planning instructions for the migration task'),
    versionControlSetupInstructions: zod_1.z
        .string()
        .describe('The version control setup instructions for the migration task'),
    generalExecutionInstructions: zod_1.z
        .string()
        .describe('The general execution workflow instructions for the migration task'),
    finalSummaryInstructions: zod_1.z
        .string()
        .describe('The final summary instructions for the migration task'),
    validations: zod_1.z
        .object({
        validationInstructions: zod_1.z
            .string()
            .describe('The general validation instructions for the migration task'),
        stages: zod_1.z.array(exports.validationStage),
    })
        .describe('The validation stages for the migration task'),
});
