// app-modernization-mcp-server/dist/kb/metadata.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataGetter = exports.ReservedSystemMetadata = exports.AcknowledgedSystemMetadata = exports.AuthorMicrosoft = void 0;
const path_1 = __importDefault(require("path"));
const utils_1 = require("../utils");
exports.AuthorMicrosoft = 'microsoft';
var AcknowledgedSystemMetadata;
(function (AcknowledgedSystemMetadata) {
    AcknowledgedSystemMetadata["Group"] = "group";
    AcknowledgedSystemMetadata["Author"] = "author";
    AcknowledgedSystemMetadata["Tags"] = "tags";
    AcknowledgedSystemMetadata["DocType"] = "docType";
})(AcknowledgedSystemMetadata || (exports.AcknowledgedSystemMetadata = AcknowledgedSystemMetadata = {}));
var ReservedSystemMetadata;
(function (ReservedSystemMetadata) {
    ReservedSystemMetadata["Source"] = "source";
    ReservedSystemMetadata["LoadedFrom"] = "loadedFrom";
    ReservedSystemMetadata["ParentMetadata"] = "parentMetadata";
    ReservedSystemMetadata["FormulaMetadata"] = "formulaMetadata";
})(ReservedSystemMetadata || (exports.ReservedSystemMetadata = ReservedSystemMetadata = {}));
class MetadataGetter {
}
exports.MetadataGetter = MetadataGetter;
_a = MetadataGetter;
MetadataGetter.getSource = (obj) => {
    if (!obj.metadata || !obj.metadata[ReservedSystemMetadata.Source]) {
        return '';
    }
    return obj.metadata[ReservedSystemMetadata.Source];
};
MetadataGetter.getLoadedFrom = (obj) => {
    if (!obj.metadata || !obj.metadata[ReservedSystemMetadata.LoadedFrom]) {
        return '';
    }
    return obj.metadata[ReservedSystemMetadata.LoadedFrom];
};
MetadataGetter.getRelativePath = (obj, includeBaseFolder = false) => {
    if (!obj.metadata || !obj.metadata[ReservedSystemMetadata.Source]) {
        return '';
    }
    const source = _a.getSource(obj);
    const loadedFrom = _a.getLoadedFrom(obj);
    if (!loadedFrom) {
        return source;
    }
    return includeBaseFolder
        ? (0, utils_1.joinToForwardSlash)(path_1.default.basename(loadedFrom), (0, utils_1.relative)(loadedFrom, source))
        : (0, utils_1.relative)(loadedFrom, source);
};
MetadataGetter.getGroup = (obj) => {
    if (!obj.metadata || !obj.metadata[AcknowledgedSystemMetadata.Group]) {
        return undefined;
    }
    return obj.metadata[AcknowledgedSystemMetadata.Group];
};
MetadataGetter.getFullGroupId = (obj) => {
    const { group, parentMetadata } = obj.metadata;
    if (!group) {
        return path_1.default.dirname(_a.getRelativePath(obj));
    }
    return _a.formatGroupId(group, parentMetadata === null || parentMetadata === void 0 ? void 0 : parentMetadata.group);
};
MetadataGetter.formatGroupId = (group, parentGroup) => {
    if (!group) {
        return '';
    }
    return parentGroup ? `${parentGroup.id}/${group.id}` : group.id;
};
MetadataGetter.getAuthor = (obj) => {
    if (!obj.metadata || !obj.metadata[AcknowledgedSystemMetadata.Author]) {
        return '';
    }
    return obj.metadata[AcknowledgedSystemMetadata.Author];
};
MetadataGetter.getTags = (obj) => {
    if (!obj.metadata || !obj.metadata[AcknowledgedSystemMetadata.Tags]) {
        return [];
    }
    return obj.metadata[AcknowledgedSystemMetadata.Tags];
};
MetadataGetter.isMicrosoft = (e) => {
    return e.metadata && e.metadata[AcknowledgedSystemMetadata.Author] === exports.AuthorMicrosoft;
};
MetadataGetter.isCustom = (e) => (0, utils_1.not)(_a.isMicrosoft)(e);
