// app-modernization-mcp-server/dist/javabuild/GradleBuildTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GradleBuildTool = void 0;
const Environment_1 = require("./context/Environment");
const JavaBuildTool_1 = require("./JavaBuildTool");
const testreport_1 = require("./test/testreport");
const gradle_1 = require("./utils/gradle");
class GradleBuildTool extends JavaBuildTool_1.JavaBuildTool {
    constructor(buildCmd = 'gradle', repoPath, env = Environment_1.Environment.Local, cmdInfo, logger, signal) {
        super(repoPath, env);
        this.buildCmd = buildCmd;
        this.env = env;
        this.cmdInfo = cmdInfo;
        this.logger = logger;
        this.signal = signal;
    }
    async getDirectDependencies() {
        const logger = this.logger || console;
        return (0, gradle_1.gradleDirectDependencies)(this.repoPath, this.buildCmd, this.env, logger);
    }
    async compile() {
        var _a, _b;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(`[GradleBuildTool] Getting build errors of ${this.repoPath} using Gradle`);
        const cmdArgs = [
            'clean',
            'compileJava',
            'compileTestJava',
            '--continue',
            '--parallel',
            '--quiet',
            '--no-daemon',
            '-Dfile.encoding=UTF-8',
            '-Dossindex.failOnError=false',
            '-Dossindex.skip=true',
            '-Dlicense.skip=true',
            '-Dcheckstyle.skip=true',
        ];
        const option = {
            logger: this.logger,
            exitOnError: false,
            throwIfFail: false,
            signal: this.signal,
            includeStdErrInStdOut: true,
        };
        const raw = await (0, gradle_1.runGradleCmd)(this.repoPath, this.buildCmd, cmdArgs, this.env, this.logger, this.cmdInfo, option);
        (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(`[GradleBuildTool] parsing compilation errors.`);
        return await (0, gradle_1.parseGradleOutput)(this.repoPath, raw);
    }
    async test() {
        var _a, _b, _c;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(`[GradleBuildTool] Running test cases of ${this.repoPath} using Gradle`);
        const cmdArgs = [
            'clean',
            'test',
            '--continue',
            '--parallel',
            '--quiet',
            '--no-daemon',
            '-Dfile.encoding=UTF-8',
            '-Dossindex.failOnError=false',
            '-Dossindex.skip=true',
            '-Dlicense.skip=true',
            '-Dcheckstyle.skip=true',
        ];
        const option = {
            logger: this.logger,
            exitOnError: false,
            throwIfFail: false,
            signal: this.signal,
            includeStdErrInStdOut: true,
        };
        const raw = await (0, gradle_1.runGradleCmd)(this.repoPath, this.buildCmd, cmdArgs, this.env, this.logger, this.cmdInfo, option);
        const compileErrors = (await (0, gradle_1.parseGradleOutput)(this.repoPath, raw)).details || [];
        (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(`[GradleBuildTool] loading surefire reports.`);
        const reports = await (0, testreport_1.getSurefireReports)(this.repoPath);
        (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info(`[GradleBuildTool] summarizing surefire reports.`);
        const result = (0, testreport_1.summarizeSurefireReports)(reports);
        result.compileErrors = compileErrors;
        return result;
    }
}
exports.GradleBuildTool = GradleBuildTool;
